PR: Add test that terminate causes EOFError

blink1073 opened the pull request:
Needed for https://github.com/jupyter/terminado/pull/142

Commit:
test that terminate causes EOFError

try with sigterm

try with sigkill too

update tests

add loop test

add explicit EOF

another attempt

cleanup

try again

don't skip the test

skip control tests for now

use new event loop

reinstate control tests

fix handling of event loop

follow docs

try skipping the loop test

try conpty only

remove more winpty

see if removing the extra read works

try without extra write and no loop test

back to original code

try winpty only

timeout after 15 minutes

try with conpty only

remove eof handling

try without capture

run pytest directly

pin pytest

fix shell quoting

use 10 minute timeout

blink1073 commented:
Okay, this caught a bug where we were using `SIGKILL`, but it doesn't yet explain what I'm seeing in terminado.  Will leave this open while I'm still debugging.

blink1073 commented:
Okay, I think what happened is https://github.com/spyder-ide/pywinpty/commit/f105bcc4b15dfe3c89d2f047d9373c14ecc7ed3c removed the ability to detect EOF passively using a `select` listening on `PtyProcess.fd`.   When we manually `read` as I've done in the tests, we get the EOF condition.  I need to make a test that uses a `select` loop that fails, and then fix the behavior.

blink1073 commented:
Note to self: we need to use [`loop.add_reader`](https://docs.python.org/3/library/asyncio-eventloop.html#watch-a-file-descriptor-for-read-events) with `asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())` to simulate how it is handled by `terminado`.

blink1073 commented:
I think `ConPTY` doesn't like reading after EOF.  Next thing I'll try is disabling that read, but only when using the `ConPTY` backend, since it might result in lost data.

blink1073 commented:
@andfoy I think `pytest` 7 made the tests in `master` start failing, I kicked the most recent [job](https://github.com/spyder-ide/pywinpty/actions/runs/1791280473) and it is also timing out with `ConPTY` and `test_intr()`.  I'm trying to debug now.

andfoy commented:
Thanks @blink1073 for tracking this, what happens if pytest is ran with `-s`? Stream redirection is difficult to track on Windows

blink1073 commented:
I tried that, and running `pytest` directly.  I'll try pinning pytest next.

blink1073 commented:
Well, it isn't pytest...

andfoy commented:
Maybe getting access into any CI worker via RDP may give us more information: https://github.com/spyder-ide/pywinpty/blob/b499f20711879bee8b0edcf00919fc78f09e51bc/.github/workflows/windows_build.yml#L55

You need an NGROK account to update this token https://github.com/spyder-ide/pywinpty/blob/b499f20711879bee8b0edcf00919fc78f09e51bc/.github/workflows/windows_build.yml#L64

blink1073 commented:
I don't have the bandwidth to debug ConPTY, my only aim is to get terminado working with WinPTY again.

andfoy commented:
You could skip the tests for conpty entirely and see if it passes

blink1073 commented:
Yep, I did that and all tests pass with `WinPTY`.

andfoy commented:
You could set `PYWINPTY_BACKEND=1`

blink1073 commented:
My vote would be to skip the specific stalling `ConPTY` tests, cut a release that fixes `terminado`, and leave an open issue for full `ConPTY` support.

Commit:
xfail on some tests with conpty

fix backend parameter

add another skip

add another skip

revert changes to test running

remove pytest pin

blink1073 commented:
Okay, ready for review.  Once this is merged I'll kick the dependabot PRs, and open an issue tracking the ConPTY regression.

ccordoba12 commented:
@blink1073, could you squash your testing commits and leave the ones that actually implemented the changes necessary for terminado? Thanks!

andfoy commented on .github/workflows/windows_build.yml:
```diff
@@ -45,13 +46,13 @@ jobs:
               run: maturin develop
             - name: Run tests
               shell: pwsh
-              run: python runtests.py
```
Can we revert to the script here?

blink1073 replied:
Done, and reverted other pytest-related changes

blink1073 commented:
> could you squash your testing commits and leave the ones that actually implemented the changes necessary for terminado

Done!

andfoy commented:
Thanks @blink1073, I'll draft a new release once I merge the dependency PRs

ccordoba12 commented:
> Done!

Thanks!

And also thank you for your hard work on solving this problem!
